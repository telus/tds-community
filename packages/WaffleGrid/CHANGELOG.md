# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.2.0](https://github.com/telus/tds-community/compare/@tds/community-waffle-grid@0.1.1...@tds/community-waffle-grid@0.2.0) (2021-09-21)


### Features

* **community-waffle-grid:** react 17 and styled components 5 ([59dfd9a](https://github.com/telus/tds-community/commit/59dfd9aefc5046ce5c3400be0658d8bbe465a7fd))





## [0.1.1](https://github.com/telus/tds-community/compare/@tds/community-waffle-grid@0.1.0...@tds/community-waffle-grid@0.1.1) (2020-09-09)


### Bug Fixes

* **community-waffle-grid:** update util-helpers version ([c1b9cd8](https://github.com/telus/tds-community/commit/c1b9cd8d588d99bef82278924b86a73809600b94))





# 0.1.0 (2020-07-13)


### Features

* **community-waffle-grid:** add new waffle grid component ([97c03e3](https://github.com/telus/tds-community/commit/97c03e3155547f03ec79be773d90fb72b3d81447))
