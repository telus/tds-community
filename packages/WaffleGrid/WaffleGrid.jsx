import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Box from '@tds/core-box'
import Link from '@tds/core-link'
import Image from '@tds/core-image'
import Paragraph from '@tds/core-paragraph'
import { media } from '@tds/core-responsive'
import { colorGreyGainsboro } from '@tds/core-colours'

import { safeRest } from '@tds/util-helpers'

const StyledLogoRow = styled('div')({
  display: 'flex',
  flexWrap: 'wrap',
  justifyContent: 'flex-start',
  ...media.from('lg').css({
    'div:not(:last-child)': {
      borderRight: `1px solid ${colorGreyGainsboro}`,
    },
    'div:nth-of-type(6n)': {
      borderRight: '0px',
    },
  }),
  ...media
    .from('sm')
    .until('lg')
    .css({
      '> div:nth-of-type(1):nth-of-last-child(3)': {
        borderBottom: '5px solid black',
      },
      'div:nth-of-type(3n - 1)': {
        borderLeft: `1px solid ${colorGreyGainsboro}`,
        borderRight: `1px solid ${colorGreyGainsboro}`,
      },
      'div:nth-of-type(3n - 1):last-child': {
        borderLeft: `1px solid ${colorGreyGainsboro}`,
      },
      '> div:not(:last-child)': {
        borderBottom: `1px solid ${colorGreyGainsboro}`,
      },
      'div:nth-last-of-type(-n+3):nth-of-type(3n) ~ *': {
        borderBottom: `1px solid ${colorGreyGainsboro}`,
        '&:last-child': {
          borderRight: '0px',
        },
      },
      'div:nth-last-of-type(-n+4):nth-of-type(3n) ~ *': {
        borderBottom: '0px',
      },
    }),
  ...media.until('sm').css({
    'div:nth-of-type(even)': {
      borderLeft: `1px solid ${colorGreyGainsboro}`,
    },
    '> div:not(:last-child)': {
      borderBottom: `1px solid ${colorGreyGainsboro}`,
    },
    '> div:nth-last-of-type(-n+2):not(:nth-of-type(even))': {
      borderBottom: '0px',
    },
  }),
})

const StyledLogoWrapper = styled('div')({
  flexDirection: 'column',
  ...media.until('sm').css({
    flex: '0 1 50%',
    width: '50%',
    alignItems: 'center',
    paddingBottom: '16px',
    '~ div:nth-of-type(n + 3)': {
      paddingTop: '16px',
    },
    '~ div:nth-last-of-type(-n + 2)': {
      paddingBottom: '0px',
    },
  }),
  ...media
    .from('sm')
    .until('lg')
    .css({
      flex: '0 1 33%',
      width: '33%',
      paddingBottom: '16px',
      '~ :nth-of-type(n + 4)': {
        paddingTop: '16px',
        paddingBottom: '0px',
      },
    }),
  ...media.from('lg').css({
    flex: '0 1 16%',
    width: '16%',
  }),
})

const StyledCentre = styled('div')({
  flexDirection: 'column',
  alignItems: 'center',
  textAlign: 'center',
})

/**
 * The WaffleGrid is used to show items in a waffle like manner with borders surrounding the element
 * @version ./package.json
 * @visibleName WaffleGrid (beta)
 */
const WaffleGrid = ({ items, ...rest }) => (
  <Box between={5} {...safeRest(rest)}>
    <StyledCentre>
      <StyledLogoRow inline between={0}>
        {items.map(child => (
          <StyledLogoWrapper key={child}>
            <StyledCentre>
              <Link href={child.href}>
                <Image src={child.image} alt={child.imageAltText} width={56} height={56} />
                <Paragraph align="center">{child.text}</Paragraph>
              </Link>
            </StyledCentre>
          </StyledLogoWrapper>
        ))}
      </StyledLogoRow>
    </StyledCentre>
  </Box>
)

WaffleGrid.propTypes = {
  /**
   * The image and the link to display. `items` should be an array of objects with the following keys:
   */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * The src attribute for the HTML img element
       */
      image: PropTypes.string,
      /**
       * The alt attribute for the HTML img element
       */
      imageAltText: PropTypes.string,
      /**
       * The text displayed under the image
       */
      text: PropTypes.string,
      /**
       * Target URL
       */
      href: PropTypes.string,
    })
  ).isRequired,
}

export default WaffleGrid
