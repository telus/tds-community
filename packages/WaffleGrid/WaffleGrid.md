### Usage criteria

- Use when you want an image and a text displayed under the image, as a link, to have a separating border either on the right or left and/or under the image/text
- Desktop/large view will have all the items in a single row
- Tablet/medium view will have the `3` items in a row
- Mobile/small view will have `2` items in a row

### Future iterations

- Handle `paddingBottom` on six or more items for `tablet` view
- Handle `borderBottom` and `paddingBottom`/`paddingTop` on 6 or more items in `desktop` view
- Handle the extra `borderBottom` and `borderRight` in `tablet` view when there are only two or three items
- Consider what should happen given a very long text to display
- Consider accepting "node" properties to be flexible to display more than just a link

```jsx
<WaffleGrid
  items={[
    {
      image:
        'https://images.ctfassets.net/3cqlnin176yn/6fbpuI7xGEQwM0sueuEMmy/8768c289e22006227aa1ebf64bd05cf4/stevie.jpg',
      imageAltText: 'The image alt text',
      href: '//telus.com',
      text: 'TELUS',
    },
    {
      image:
        'https://images.ctfassets.net/3cqlnin176yn/6fbpuI7xGEQwM0sueuEMmy/8768c289e22006227aa1ebf64bd05cf4/stevie.jpg',
      imageAltText: 'The image alt text',
      href: '//telus.com',
      text: 'TELUS',
    },
    {
      image:
        'https://images.ctfassets.net/3cqlnin176yn/6fbpuI7xGEQwM0sueuEMmy/8768c289e22006227aa1ebf64bd05cf4/stevie.jpg',
      imageAltText: 'The image alt text',
      href: '//telus.com',
      text: 'TELUS',
    },
    {
      image:
        'https://images.ctfassets.net/3cqlnin176yn/6fbpuI7xGEQwM0sueuEMmy/8768c289e22006227aa1ebf64bd05cf4/stevie.jpg',
      imageAltText: 'The image alt text',
      href: '//telus.com',
      text: 'TELUS',
    },
    {
      image:
        'https://images.ctfassets.net/3cqlnin176yn/6fbpuI7xGEQwM0sueuEMmy/8768c289e22006227aa1ebf64bd05cf4/stevie.jpg',
      imageAltText: 'The image alt text',
      href: '//telus.com',
      text: 'TELUS',
    },
    {
      image:
        'https://images.ctfassets.net/3cqlnin176yn/6fbpuI7xGEQwM0sueuEMmy/8768c289e22006227aa1ebf64bd05cf4/stevie.jpg',
      imageAltText: 'The image alt text',
      href: '//telus.com',
      text: 'TELUS',
    },
  ]}
/>
```
