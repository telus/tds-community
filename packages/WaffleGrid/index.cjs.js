const WaffleGrid = require('./dist/index.cjs')

module.exports = WaffleGrid
