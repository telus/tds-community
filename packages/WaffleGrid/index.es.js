import WaffleGrid from './dist/index.es'

export default WaffleGrid
