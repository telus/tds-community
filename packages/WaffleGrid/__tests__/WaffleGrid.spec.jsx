import React from 'react'
import { shallow } from 'enzyme'

import WaffleGrid from '../WaffleGrid'

describe('WaffleGrid', () => {
  const defaultProps = {
    items: [
      {
        image: 'www.image.com/cool.svg',
        imageAltText: 'Cool image',
        href: 'www.image.com',
        text: 'Cool image',
      },
    ],
  }
  const doShallow = (props = {}) => shallow(<WaffleGrid {...defaultProps} {...props} />)

  it('renders', () => {
    const waffleGrid = doShallow()

    expect(waffleGrid).toMatchSnapshot()
  })

  it('passes additional attributes to the element', () => {
    const waffleGrid = doShallow({ id: 'the-id', 'data-some-attr': 'some value' })

    expect(waffleGrid).toHaveProp('id', 'the-id')
    expect(waffleGrid).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const waffleGrid = doShallow({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(waffleGrid).not.toHaveProp('className', 'my-custom-class')
    expect(waffleGrid).not.toHaveProp('style')
  })
})
