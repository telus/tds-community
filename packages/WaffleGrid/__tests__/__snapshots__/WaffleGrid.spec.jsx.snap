// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WaffleGrid renders 1`] = `
<Box
  between={5}
  inline={false}
  tag="div"
>
  <WaffleGrid__StyledCentre>
    <WaffleGrid__StyledLogoRow
      between={0}
      inline={true}
    >
      <WaffleGrid__StyledLogoWrapper
        key="[object Object]"
      >
        <WaffleGrid__StyledCentre>
          <WithForwardedRef(Link)
            href="www.image.com"
          >
            <Image
              alt="Cool image"
              height={56}
              src="www.image.com/cool.svg"
              width={56}
            />
            <Paragraph
              align="center"
              bold={false}
              invert={false}
              size="medium"
            >
              Cool image
            </Paragraph>
          </WithForwardedRef(Link)>
        </WaffleGrid__StyledCentre>
      </WaffleGrid__StyledLogoWrapper>
    </WaffleGrid__StyledLogoRow>
  </WaffleGrid__StyledCentre>
</Box>
`;
